### DISCUSSION TITLE
setMutationDefaults and getMutationDefaults typings

### DISCUSSION THREAD

**Discussion by DarioSiroki** [TanStack/query#5955] [Ideas]
Is there any reason why these are typed as:

```ts
  setMutationDefaults(
    mutationKey: MutationKey,
    options: MutationObserverOptions<any, any, any, any>,
  ): void {
...
  }

  getMutationDefaults(
    mutationKey?: MutationKey,
  ): MutationObserverOptions<any, any, any, any> | undefined {
...  
}
```

instead of like this?

```ts
 function getMutationDefaults<TData=unknown, TError=unknown, TVariables=void, TContext=unknown>(
    mutationKey?: MutationKey,
  ): MutationObserverOptions<TData, TError, TVariables, TContext> | undefined {}
```

I'm currently using satisfies to work around this:

```ts
queryClient.setMutationDefaults(['createNoteForLocation'], {
    mutationFn: (note) => createNoteForLocation(note),
    onMutate: async (note) => {
       ....

        return { previousData, images: note.images }
    },
    onSettled: async (response, error, production) => {
     ...
    },
    onError: (error, _, context) => {
     ....
    },
} satisfies MutationObserverOptions<AddNoteDetailsResponse, Error, AddNoteDetails, { previousData: LocationDetails | undefined, images: string[] }>)
```

**Comment by TkDodo** [comment]
for `getMutationDefaults`, there would be no way to infer those generics. You would have to do:

```
queryClient.getMutationDefaults<Generics, Go, Here>(...)
```

and that's pretty much the same as using a type assertion. Also, in v5, defaults are properly merged, so if you have multiple defaults for multiple layers of keys, there is no way to know on type level what you will get. So it's an unsafe assumption anyways.

for `setMutationDefaults`, it only makes sense when you define the `mutationFn` as default, because generics are inferred from there. I guess I'd accept a PR to change those types (for v5, `beta` branch please)

**Reply by DarioSiroki** [reply]
Oops, brainfart, I thought the second example was for `setQueryDefaults` and I actually pasted in the `getMutationDefaults`. `setQueryDefaults` could use the same improvement as `setMutationDefaults`. 

I'll open a PR over the weekend, thanks

**Comment by toppsdown** [comment]
Hey there, I just stumbled across this same situation where the default mutation wasn't inferring types in the onMutate/Success/Error functions.

Using "satisfies" works. Is that the recommended way to add typing to setMutationDefaults?
